File: python/snake.py
import pygame
import sys
import copy
import random

pygame.init()

# Configuration
width = 500
height = 500
screen = pygame.display.set_mode((width, height))
cellSize = 15;
cellsX = round(width/cellSize);
cellsY = round(height/cellSize);
snake = None
food = None
gameloop = True
fps = 20

# set default drawing colors
color = (0,0,0)
bgColor = (255,255,255)


class SNAKE:
    def __init__(self):
        self.elements = []
        self.direction = 'right'
        self.length = 5

    def create(self):
        self.elements = []
        i = self.length
        while i != 0:
            self.elements.append({'x': i, 'y':0})
            i -= 1
    def collidesWith(self, otherElement):
			for element in self.elements:
				if (element['x'] == otherElement['x'] and
					element['y'] == otherElement['y']):
					return True
			return False;

    def render(self):
        for element in self.elements:
            draw(element)

    def move(self):
        global gameloop
        elementToMove = copy.deepcopy(self.elements)
        elementToMove = elementToMove[0]
        if self.direction == 'right':
            elementToMove['x'] = elementToMove['x'] + 1
        elif self.direction == 'left':
            elementToMove['x'] = elementToMove['x'] - 1
        elif self.direction == 'down':
            elementToMove['y'] = elementToMove['y'] + 1
        elif self.direction == 'up':
            elementToMove['y'] = elementToMove['y'] - 1

        if (elementToMove['x'] == -1 or
            elementToMove['x'] == cellsX or
            elementToMove['y'] == -1 or
            elementToMove['y'] == cellsY or
            self.collidesWith(elementToMove)):
			# restart game
			gameloop = False

        self.elements.pop()
        self.elements.insert(0, elementToMove)

class FOOD:
    def __init__(self):
        self.position = None

    def create(self):
        self.position = {
            'x': random.randrange(0, (width - cellSize)/cellSize, 1),
            'y': random.randrange(0, (height - cellSize)/cellSize, 1)
        }

    def render(self):
        draw(self.position)

def initGame():
    global snake, food
    snake = SNAKE()
    snake.direction = 'right'
    snake.create()
    food = FOOD()
    food.create()
    score = 0


def draw(element):
    pygame.draw.rect(
        screen,
        color,
        (
            element['x'] * cellSize,
            element['y'] * cellSize,
            cellSize,
            cellSize
        ),
        0
    )

def render():
    snake.move()
    snake.render()
    food.render()

initGame()

# main game loop
while (True):
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit(); sys.exit();
        if event.type == pygame.KEYDOWN:
            if event.key == pygame.K_LEFT and snake.direction != 'right':
                snake.direction = 'left'
            if event.key == pygame.K_RIGHT and snake.direction != 'left':
                snake.direction = 'right'
            if event.key == pygame.K_DOWN and snake.direction != 'up':
                snake.direction = 'down'
            if event.key == pygame.K_UP and snake.direction != 'down':
                snake.direction = 'up'

    # fill screen with default background color
    screen.fill(bgColor);

    #draw single rectangle
    if gameloop:
        render()
        pygame.display.update()
        pygame.time.wait(1000/fps)

File: js/snake.js
//Snake game
// Michał Budzyński
// Language sandbox Javascript example

var GAME = function() {
	// config & variables
	var canvas = document.getElementById('canvas');
	var ctx = canvas.getContext('2d');
	var canvasWidth = canvas.width;
	var canvasHeight = canvas.height;
	var cellSize = 15;
	var cellsX = Math.round(canvasWidth/cellSize);
	var cellsY = Math.round(canvasHeight/cellSize);
	var score;
	var gameLoop;

	var SNAKE = {
		elements: [],
		direction: 'right',
		length: 5,
		create: function(){
			this.elements = [];
			var i = this.length;
			while(i--) {
				this.elements.push({x: i, y:0});
			}
		},
		collidesWith: function(element){
			for(var i = 0; i < this.elements.length; i++) {
				if(
					this.elements[i].x == element.x &&
					this.elements[i].y == element.y
				) {
					return true;
				}
			}
			return false;
		}
	};

	var FOOD = {
		position: {},
		create: function(){
			this.position = {
				x: Math.round(Math.random()*(canvasWidth - cellSize)/cellSize),
				y: Math.round(Math.random()*(canvasHeight - cellSize)/cellSize)
			}
		}
	};

	var init = function() {
		SNAKE.direction = 'right';
		SNAKE.create();
		FOOD.create();
		score = 0;

		clearInterval(gameLoop);
		gameLoop = setInterval(render, 260);
	}

	//Lets paint the snake now
	function render() {
		// clear the canvas
		canvas.width = canvas.width;

		// copy the tail of the snake, this is the easiest way to do it in JS
		var elementToMove = JSON.parse(JSON.stringify(SNAKE.elements[0]));
		switch(SNAKE.direction) {
			case 'right':
				elementToMove.x++;
				break;
			case 'left':
				elementToMove.x--;
				break;
			case 'up':
				elementToMove.y--;
				break;
			case 'down':
				elementToMove.y++;
				break;
		};

		// game over conditions
		if(
			elementToMove.x === -1 ||
			elementToMove.x === cellsX ||
			elementToMove.y === -1 ||
			elementToMove.y === cellsY ||
			SNAKE.collidesWith(elementToMove)
		) {
			//restart game
			alert("Game Over!\nScore: " + score);
			init();
			return;
		}

		SNAKE.elements.unshift(elementToMove);

		if(
			elementToMove.x === FOOD.position.x &&
			elementToMove.y === FOOD.position.y
		) {
			// snake ate food
			score++;
			FOOD.create();
		} else {
			// snake didn't
			SNAKE.elements.pop();
		}

		SNAKE.elements.forEach(function(element){
			draw(element);
		});
		draw(FOOD.position);
		ctx.fillText('SCORE: ' + score, 5, canvasHeight - 5);
	}

	function draw(element) {
		ctx.fillRect(
			element.x * cellSize,
			element.y * cellSize,
			cellSize,
			cellSize
		);
		ctx.strokeStyle = 'white';
		ctx.strokeRect(
			element.x * cellSize,
			element.y * cellSize,
			cellSize,
			cellSize
		);
	}

	document.addEventListener('keydown', function(e) {
		var key = e.which;
		if(key === 37 && SNAKE.direction !== 'right') SNAKE.direction = 'left';
		else if(key === 38 && SNAKE.direction !== 'down') SNAKE.direction = 'up';
		else if(key === 39 && SNAKE.direction !== 'left') SNAKE.direction = 'right';
		else if(key === 40 && SNAKE.direction !== 'up') SNAKE.direction = 'down';
	});

	init();
};

GAME();
